### PR TITLE
Add temporary event listener to the Web Extensions Test API.

### PR DESCRIPTION
**Pull request by kiaraarose**
#### 2e30913e291b5fcbd40523d86da31c9fc61d8b9c
<pre>
Add temporary event listener to the Web Extensions Test API.
<a href="https://bugs.webkit.org/show_bug.cgi?id=248972">https://bugs.webkit.org/show_bug.cgi?id=248972</a>

Reviewed by Brian Weinstein and Timothy Hatcher.

This patch adds a temporary event listener to the Test API and a couple of test cases to TestWebKitAPI.

* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPITestCocoa.mm:
(WebKit::WebExtensionAPITest::testEvent):
(WebKit::WebExtensionAPITest::fireTestEvent):
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPIEvent.h:
(WebKit::WebExtensionAPIEvent::WebExtensionAPIEvent):
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPITest.h:
* Source/WebKit/WebProcess/Extensions/Interfaces/WebExtensionAPITest.idl:
* Tools/TestWebKitAPI/SourcesCocoa.txt:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIEvent.mm:
Copied from Source/WebKit/WebProcess/Extensions/API/WebExtensionAPIEvent.h.
(TestWebKitAPI::TEST):

Canonical link: <a href="https://commits.webkit.org/257797@main">https://commits.webkit.org/257797@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/00b804ec598eec01d585f31232857b368db89a4d

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/100050 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/9220 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/33131 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/109396 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/169633 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/10100 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/86627 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/92496 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/107292 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/105817 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/7635 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/90919 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/34347 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/89542 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/22312 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [⏳ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/API-Tests-GTK-EWS "Waiting to run tests") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/81/builds/3008 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/23827 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/86592 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/83/builds/419 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/2968 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/46196 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/29031 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/9101 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/62/builds/43308 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/89471 "Built successfully") | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/79/builds/4818 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/20003 "Passed tests") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/2755 "Built successfully and passed tests") | | | |

### CODE REVIEW

File: Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIEvent.mm [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+
+#if ENABLE(WK_WEB_EXTENSIONS)
+
+#import "WebExtensionUtilities.h"
+
+namespace TestWebKitAPI {
+
+TEST(WKWebExtensionAPIEvent, EventListenerAdded)
+{
+    auto *backgroundScript = Util::constructScript(@[
```
Comment (kiaraarose) [inline Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIEvent.mm:36]: Wasn't able to verify that these test pass locally, but I'll wait to see if the tests on the bot pass. Do these test cases look ok though?

File: Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPITestCocoa.mm [hunk 2]
```diff
@@ -200,6 +201,19 @@
     testEvent.invokeListenersWithArgument(@{ @"url": urlString }, targetURL);
 }
 
+WebExtensionAPIEvent& WebExtensionAPITest::testEvent()
+{
+    if (!m_event)
+        m_event = WebExtensionAPIEvent::create(forMainWorld(), runtime(), extensionContext(), WebExtensionEventListenerType::ActionOnClicked);
+
```
Comment (b-weinstein) [inline Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPITestCocoa.mm:208]: Super-nit: No need for the blank line here.

File: Source/WebKit/WebProcess/Extensions/API/WebExtensionAPITest.h [hunk 3]
```diff
@@ -62,10 +63,13 @@ class WebExtensionAPITest : public WebExtensionAPIObject, public JSWebExtensionW
     void assertThrows(JSContextRef, JSValue *function, JSValue *expectedError, NSString *message);
 
     WebExtensionAPIWebNavigationEvent& testWebNavigationEvent();
+    WebExtensionAPIEvent& testEvent();
     void fireTestWebNavigationEvent(NSString *urlString);
+    void fireTestEvent();
```
Comment (b-weinstein) [inline Source/WebKit/WebProcess/Extensions/API/WebExtensionAPITest.h:69]: Nit: I think the two methods you are adding could be paired together.

File: Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIEvent.mm (hunk 1 above)
Comment (b-weinstein) [reply Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIEvent.mm:36]: I think you could do all of these in one test.

1) When starting, check that hasListener is false
2) Add the listener
3) Check that hasListener is true
4) Remove the listener
5) Check that hasListener is false

But what you have here is okay too, just has some overhead.

I think there could also be a test for firing the event.

**Review by b-weinstein — APPROVED** [review]

File: Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIEvent.mm [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+    auto *manifest = @{ @"manifest_version": @2, @"background": @{ @"scripts": @[ @"background.js" ], @"type": @"module", @"persistent": @NO } };
+
+    Util::loadAndRunExtension(manifest, @{ @"background.js": backgroundScript });
+}
+
+TEST(WKWebExtensionAPIEvent, EventListenerRemoved)
+{
+    auto *backgroundScript = Util::constructScript(@[
+        // Setup
+        @"function test() { console.log(\"test\") }",
```
Comment (xeenon) [inline Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIEvent.mm]: Could use `browser.test.log()` to see the output during the test.

File: Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIEvent.mm [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+    auto *manifest = @{ @"manifest_version": @2, @"background": @{ @"scripts": @[ @"background.js" ], @"type": @"module", @"persistent": @NO } };
+
+    Util::loadAndRunExtension(manifest, @{ @"background.js": backgroundScript });
+}
+
+TEST(WKWebExtensionAPIEvent, NoEventListenerAdded)
+{
+    auto *backgroundScript = Util::constructScript(@[
+        // Setup
+        @"function test() { console.log(\"test\") }",
```
Comment (xeenon) [inline Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIEvent.mm]: Ditto.

**Review by xeenon — APPROVED** [review]

File: Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIEvent.mm [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+
+#import "WebExtensionUtilities.h"
+
+namespace TestWebKitAPI {
+
+TEST(WKWebExtensionAPIEvent, TestEventListener)
+{
+    auto *backgroundScript = Util::constructScript(@[
+        // Setup
+        @"function test() { browser.test.log(\"test\") }",
```
Comment (b-weinstein) [inline Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIEvent.mm]: Nit: Before adding the listener, you could have:

`browser.test.assertFalse(browser.test.hasListener(test))",`

File: Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIEvent.mm [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+#import "WebExtensionUtilities.h"
+
+namespace TestWebKitAPI {
+
+TEST(WKWebExtensionAPIEvent, TestEventListener)
+{
+    auto *backgroundScript = Util::constructScript(@[
+        // Setup
+        @"browser.test.assertFalse(browser.test.hasListener(test))",
+        @"function test() { browser.test.log(\"test\") }",
```
Comment (xeenon) [inline Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIEvent.mm]: This should go first, but JavaScript allows it in any order.

File: Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIEvent.mm [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+        @"browser.test.testEvent.addListener(test)",
+
+        // Normal Cases
+        @"browser.test.assertTrue(browser.test.hasListener(test))",
+        @"browser.test.testEvent.removeListener(test)",
+        @"browser.test.assertFalse(browser.test.hasListener(test))",
+
+        // Finish
```
Comment (xeenon) [inline Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIEvent.mm]: Did you want to add a `browser.test.fireEvent()` call here?

**Review by xeenon — APPROVED** [review]

**Pull request merged**